src/compiler/checker.ts

```diff
@@ -24005,6 +23995,17 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         if (inInferTypeForHomomorphicMappedType) {
             return undefined;
         }
+        if (constraint.type.flags & TypeFlags.IndexedAccess
+            && (constraint.type as IndexedAccessType).objectType.flags & TypeFlags.TypeParameter
+            && (constraint.type as IndexedAccessType).indexType.flags & TypeFlags.TypeParameter) {
+            // A reverse mapping of `{[K in keyof T[K_1]]: T[K_1]}` is the same as that of `{[K in keyof T]: T}`, since all we care about is
+            // inferring to the "type parameter" (or indexed access) shared by the constraint and template. So, to reduce the number of
+            // type identities produced, we simplify such indexed access occurences
+            const newTypeParam = (constraint.type as IndexedAccessType).objectType;
+            const newMappedType = replaceIndexedAccess(target, constraint.type as ReplaceableIndexedAccessType, newTypeParam);
+            target = newMappedType as MappedType;
+            constraint = getIndexType(newTypeParam) as IndexType;

```

Comment:
By moving this logic from reverse mapped type property construction forward into reverse mapped type construction, we can cut down on the number of reverse mapped type identities (and property symbols) we make - this is critical to actually being able to block the recursion, since it makes a generative recursion into a self recursion.